CONTRIBUTING.md

```diff
@@ -0,0 +1,77 @@
+
+## Coding conventions
+
+* Ensure to document all exported functions, methods, and structs, according to the [Golang documentation guidelines](https://blog.golang.org/godoc).
+* Avoid use of gotos as they lead to spaghetti code.
+* No else in the if statement, should be the default. Try to always limit the amount of indentation used, as it might make it harder to read code.
+* New code always needs to be tested. Pull requests will generally not be merged if there is no proper test coverage of newly added code.
+* If you fix a bug which was hard to find, create/add to the `gotchas.md` file in the package. This helps other developers catch the bug much quicker if it pops up again.
+* If you are adding new code, please include proper documentation outside of code comments where it makes sense. Ideally, for adding new features, it helps to at the very minimum add a `README.md` which explains purpose, inner workings, etc. In some cases, it may also be helpful to include UML diagrams which depict the new feature/component (sequence diagrams, component diagrams and workflows are often good choices).
+* Use the `context` package to manage goroutine lifetimes. For good examples, look at how the `Chain` and `Mempool` are managed using contexts.
+
+### Performance
+
+* Wherever possible, when allocating slices, try to allocate the entire slice in one go, and then assigning to the indices (or using `copy` where applicable), instead of creating an empty slice and then calling `append`. This prevents an unnecessary amount of dynamic memory allocations.
+* When creating a channel which sole purpose is to send signals, and not data, use a `chan struct{}`. The empty struct in Go takes up 0 bytes of space, and is thus the most efficient way of sending signals.
+* When clearing a map, use a `range` loop, and delete every key. This is more efficient than replacing the map with a new instance, as the GC does not have to clean this up later.
+
+### Committing
+
+* Rebase your commits generously. Avoid having to open pull requests with a massive amount of commits, which all essentially work towards the same issue.
+* If possible, try to include all of the code, tests and documentation for one issue, into one commit.
+* The commit title should be in imperative form. Examples:
+  * Implement ...
+  * Add ...
+  * Change ...
+* The commit title should only state what you did. This helps keep commit titles short and easy to find.
+* The commit description should explain why you did it. Most times, referring to the issue itself is enough. Please do so with keywords (`fixes`, `resolves` etc.) in order to automatically link issues to commits. This helps to retrace commits and it simplifies opening pull requests.
+* Do not elaborate on how you did what you did - the code should be self-explanatory in this case. If the code is not self-explanatory enough (due to performance shortcuts or the sort), please explain this in comments, rather than in the commit.
+
```

Comment:
We should add a recommended line size for commit descriptions (I think it is around 80 characters)